### PR TITLE
Second feature. Card list and card details

### PR DESCRIPTION
**Pull request by FacuCristobalGlobant**
Second feature completed, the app shows a list of cards obtained from a json file and clicking one shows its details in a new view using navigator.

https://user-images.githubusercontent.com/109153713/180835627-2440f9b3-b685-4ec9-8062-37ba9a7dde3e.mp4

### CODE REVIEW

File: lib/src/repository/card_list_repo.dart [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import 'dart:convert';
+import 'package:flutter/services.dart';
+import 'package:hearthstoneapp/src/model/hearthstone_card.dart';
```
Comment (Juliana-Dimeglio) [inline lib/src/repository/card_list_repo.dart]: This could be a relative import

File: lib/widgets/card.dart [hunk 2]
```diff
@@ -0,0 +1,131 @@
+import 'package:flutter/material.dart';
+import 'package:hearthstoneapp/src/model/hearthstone_card.dart';
+import 'package:hearthstoneapp/utils/constants.dart';
+import 'package:hearthstoneapp/utils/palette.dart';
+import 'package:hearthstoneapp/widgets/card_property.dart';
+import 'package:hearthstoneapp/widgets/like_counter.dart';
+import 'package:hearthstoneapp/widgets/outlined_text.dart';
```
Comment (Juliana-Dimeglio) [inline lib/widgets/card.dart:8]: Relative imports

File: lib/widgets/card_property.dart [hunk 3]
```diff
@@ -1,27 +1,59 @@
 import 'package:flutter/material.dart';
+import 'package:hearthstoneapp/utils/constants.dart';
```
Comment (Juliana-Dimeglio) [inline lib/widgets/card_property.dart:2]: Relative import

File: lib/widgets/list_of_cards.dart [hunk 4]
```diff
@@ -0,0 +1,115 @@
+import 'package:flutter/material.dart';
+import 'package:hearthstoneapp/src/model/hearthstone_card.dart';
+import 'package:hearthstoneapp/src/repository/card_list_repo.dart';
+import 'package:hearthstoneapp/utils/constants.dart';
+import 'package:hearthstoneapp/utils/palette.dart';
+import 'package:hearthstoneapp/widgets/card.dart';
```
Comment (Juliana-Dimeglio) [inline lib/widgets/list_of_cards.dart:7]: Relative imports

File: lib/widgets/outlined_text.dart [hunk 5]
```diff
@@ -1,4 +1,5 @@
 import 'package:flutter/material.dart';
+import 'package:hearthstoneapp/utils/constants.dart';
```
Comment (Juliana-Dimeglio) [inline lib/widgets/outlined_text.dart:2]: Relative import

File: lib/src/repository/card_list_repo.dart [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+abstract class CardListRepo {
```
Comment (CIFUM007) [inline lib/src/repository/card_list_repo.dart]: why did you defined this class as abstract?

File: lib/widgets/card.dart [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+class CardView extends StatelessWidget {
+  final HearthstoneCard card;
+
+  const CardView({Key? key, required this.card}) : super(key: key);
```
Comment (CIFUM007) [inline lib/widgets/card.dart]: trailing comma missing after card

File: lib/src/repository/card_list_repo.dart (hunk 6 above)
Comment (CIFUM007) [reply lib/src/repository/card_list_repo.dart]: also, you don't need to shorten the name of the class, repository is okey, for the name of the class and for the file, rename them pliz.
It's a good practice to have complete and understandable names, than make assumptions of what are you trying to say

File: lib/src/repository/card_list_repo.dart (hunk 1 above)
Comment (FacuCristobalGlobant) [reply lib/src/repository/card_list_repo.dart]: I can't use relative imports for some unknown reason, Once we discover how i'll fix them.

File: lib/src/repository/card_list_repo.dart (hunk 6 above)
Comment (FacuCristobalGlobant) [reply lib/src/repository/card_list_repo.dart]: I've applied those changes, the class is no longer abstract and I've changed the names of the class and the file.

File: lib/widgets/card.dart (hunk 7 above)
Comment (FacuCristobalGlobant) [reply lib/widgets/card.dart]: Added

File: lib/src/model/hearthstone_card.dart [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    required this.playerClass,
+    required this.name,
+    required this.type,
+    this.text,
+    this.health,
+    this.cost,
+    this.attack,
+    required this.locale,
+    this.mechanics,
+    String img = "assets/images/EX1_572.png",
```
Comment (FranciscoGaraventa) [inline lib/src/model/hearthstone_card.dart]: Add this asset route to constant file

File: lib/src/repository/card_list_repository.dart [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    data.forEach((card) {
+      HearthstoneCard currentCard =
+          HearthstoneCard.fromJson(card as Map<String, dynamic>);
+      listOfCards.add(currentCard);
+    });
+
+    return listOfCards;
+  }
+
+  CardListRepository();
```
Comment (FranciscoGaraventa) [inline lib/src/repository/card_list_repository.dart]: Do you need this constructor?

File: lib/widgets/card.dart [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+                      const Text(
+                        Constants.cardAbilityLabel,
+                        style: TextStyle(
+                          fontSize: Constants.abilityLabelFontSize,
+                          fontWeight: FontWeight.bold,
+                          color: Colors.black87,
+                        ),
+                      ),
+                      Text(
+                        card.text ?? "",
```
Comment (FranciscoGaraventa) [inline lib/widgets/card.dart]: You can add "" as emptyString to constant file

File: lib/widgets/card_property.dart [hunk 11]
```diff
@@ -1,27 +1,59 @@
...
-        fontWeight: weight,
+    return Visibility(
+      visible: visibility,
+      child: Padding(
+        padding: Constants.verticalSmallPadding,
+        child: Row(
+          mainAxisAlignment: MainAxisAlignment.spaceBetween,
+          children: [
+            Flexible(
+              flex: Constants.flex2,
```
Comment (FranciscoGaraventa) [inline lib/widgets/card_property.dart]: We dont use number to name constants, here you can change it instead to flex2 like cardPropertyFlex or something similar

File: lib/widgets/card_property.dart [hunk 12]
```diff
@@ -1,27 +1,59 @@
...
+                textAlign: TextAlign.start,
+                style: TextStyle(
+                  fontSize: fontSize,
+                  color: color,
+                  fontWeight: weight,
+                ),
+              ),
+            ),
+            Flexible(
+              flex: Constants.flex3,
```
Comment (FranciscoGaraventa) [inline lib/widgets/card_property.dart]: Same as above

File: lib/widgets/list_of_cards.dart [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+      ),
+      body: FutureBuilder(
+        future: listOfCards,
+        builder: (
+          context,
+          AsyncSnapshot<List<HearthstoneCard>> snapshot,
+        ) {
+          List<Widget> children = [];
+          if (snapshot.hasData) {
+            snapshot.data?.forEach((element) {
```
Comment (FranciscoGaraventa) [inline lib/widgets/list_of_cards.dart]: You know that snapshot.data gave you a List<HearthstoneCard> so instead to name inside for each an 'element' you can name it hearthstoneCard

File: lib/widgets/list_of_cards.dart [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+                              ),
+                              Text(
+                                element.cardSet,
+                                style: Constants.iconFontTextStyle,
+                              ),
+                            ],
+                          ),
+                        ),
+                        Flexible(
+                          flex: Constants.flex3,
```
Comment (FranciscoGaraventa) [inline lib/widgets/list_of_cards.dart]: You can rename like cardListImageFlex

File: lib/src/model/hearthstone_card.dart (hunk 8 above)
Comment (FacuCristobalGlobant) [reply lib/src/model/hearthstone_card.dart]: created a constant with the name "defaultImage"

File: lib/src/repository/card_list_repository.dart (hunk 9 above)
Comment (FacuCristobalGlobant) [reply lib/src/repository/card_list_repository.dart]: I don't! removed.

File: lib/widgets/card.dart (hunk 10 above)
Comment (FacuCristobalGlobant) [reply lib/widgets/card.dart]: Added

File: lib/widgets/card_property.dart (hunk 11 above)
Comment (FacuCristobalGlobant) [reply lib/widgets/card_property.dart]: Changed them to propertyCardSmall, ...Medium and ...Large respectively

File: lib/widgets/card_property.dart (hunk 12 above)
Comment (FacuCristobalGlobant) [reply lib/widgets/card_property.dart]: Fixed

File: lib/widgets/list_of_cards.dart (hunk 13 above)
Comment (FacuCristobalGlobant) [reply lib/widgets/list_of_cards.dart]: Done!

File: lib/widgets/card_property.dart (hunk 11 above)
Comment (FacuCristobalGlobant) [reply lib/widgets/card_property.dart]: To be able to use the values in other places I created a CustomFlex class with the constant values for small, medium and large.

File: lib/widgets/list_of_cards.dart (hunk 14 above)
Comment (FacuCristobalGlobant) [reply lib/widgets/list_of_cards.dart]: Changed it to CustomFlex.medium.

**Review by Juliana-Dimeglio — APPROVED** [review]

**Review by FranciscoGaraventa — APPROVED** [review]
LGTM

File: lib/widgets/list_of_cards.dart [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+import 'package:hearthstoneapp/src/model/hearthstone_card.dart';
+import 'package:hearthstoneapp/src/repository/card_list_repository.dart';
+import 'package:hearthstoneapp/utils/constants.dart';
+import 'package:hearthstoneapp/utils/custom_flex.dart';
+import 'package:hearthstoneapp/utils/palette.dart';
+import 'package:hearthstoneapp/widgets/card.dart';
+
+class ListOfCards extends StatelessWidget {
+  ListOfCards({Key? key}) : super(key: key);
+  final CardListRepository repo = CardListRepository();
```
Comment (CIFUM007) [inline lib/widgets/list_of_cards.dart]: remember that it's not necessary to shorten names, rename "repo" pliz

Comment (FacuCristobalGlobant) [reply lib/widgets/list_of_cards.dart]: Done!

**Pull request merged**